src/java_bytecode/java_types.cpp

```diff
@@ -614,6 +617,12 @@ std::vector<typet> java_generic_type_from_string(
   {
     size_t bound_sep=signature.find(':');
     INVARIANT(bound_sep!=std::string::npos, "No bound for type variable.");
+
+    // is bound a class?
+    // in this case the separator is '::'
```

Comment:
This seems surprising? The [spec](https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.3.4) doesn't appear to mention anything about a `::` separator?

Reply:
Ah no, my mistake. The idea is you have the identifier, then a colon, then optionally the class bound, then optionally a `:interface bound`. Ie:

`T extends Class`: `T:Class`
`T extends Interface`: `T::Interface`
`T extends Interface1 & Interface2`: `T::Interface1:Interface2`
`T extends Class & Interface1 & Interface2`: `T:Class:Interface1:Interface2`


I think the comment should make this clearer and perhaps verify that this only happens once and first.